src/middlewared/middlewared/plugins/account.py

```diff
@@ -27,50 +81,477 @@ class UserService(CRUDService):
-        if not user:
-            raise CallError(f'User {id} does not exist', errno.ENOENT)
-        user = user[0]
-
-        if 'sshpubkey' in data:
-            keysfile = f'{user["home"]}/.ssh/authorized_keys'
-            pubkey = data.pop('sshpubkey')
-            if pubkey is None:
-                if os.path.exists(keysfile):
-                    try:
-                        os.unlink(keysfile)
-                    except OSError:
-                        pass
+    @accepts(Dict(
+        'user_create',
+        Int('uid'),
+        Str('username', required=True),
+        Int('group'),
+        Bool('group_create', default=False),
+        Str('home', default='/nonexistent'),
+        Str('home_mode', default='755'),
+        Str('shell', default='/bin/csh'),
+        Str('full_name', required=True),
+        Str('email'),
+        Str('password'),
+        Bool('password_disabled', default=False),
+        Bool('locked', default=False),
+        Bool('microsoft_account', default=False),
+        Bool('sudo', default=False),
+        Str('sshpubkey'),
+        List('groups'),
+        Dict('attributes', additional_attrs=True),
+        register=True,
+    ))
+    async def do_create(self, data):
+
+        verrors = ValidationErrors()
+
+        if (
+            not data.get('group') and not data.get('group_create')
+        ) or (
+            data.get('group') is not None and data.get('group_create')
+        ):
+            verrors.add('group', f'You need to either provide a group or group_create', errno.EINVAL)
+
+        await self.__common_validation(verrors, data)
+
+        if data.get('sshpubkey') and not data['home'].startswith('/mnt'):
+            verrors.add('sshpubkey', 'Home directory is not writable, leave this blank"')
+
+        if verrors:
+            raise verrors
+
+        groups = data.pop('groups') or []
+        create = data.pop('group_create')
+
+        if create:
+            group = await self.middleware.call('group.query', [('group', '=', data['username'])])
+            if group:
+                group = group[0]
+            else:
+                group = await self.middleware.call('group.create', {'name': data['username']})
+                group = (await self.middleware.call('group.query', [('id', '=', group)]))[0]
+            data['group'] = group['id']
+        else:
+            group = await self.middleware.call('group.query', [('id', '=', data['group'])])
+            if not group:
+                raise CallError(f'Group {data["group"]} not found')
+            group = group[0]
+
+        # Is this a new directory or not? Let's not nuke existing directories,
+        # e.g. /, /root, /mnt/tank/my-dataset, etc ;).
+        new_homedir = False
+        home_mode = data.pop('home_mode')
+        if data['home'] != '/nonexistent':
+            try:
+                os.makedirs(data['home'], mode=int(home_mode, 8))
+                if os.stat(data['home']).st_dev == os.stat('/mnt').st_dev:
```

Comment:
If directory already exists, `os.makedirs` will raise exception and this check will be skipped